gotham/src/lib.rs

```diff
@@ -69,17 +88,69 @@ where
     start_with_num_threads(addr, threads, new_handler)
 }
 
-fn tcp_listener<A>(addr: A) -> (TcpListener, SocketAddr)
+fn start_with_num_threads<NH, A>(addr: A, threads: usize, new_handler: NH)
 where
+    NH: NewHandler + 'static,
     A: ToSocketAddrs,
 {
-    let addr = match addr.to_socket_addrs().map(|ref mut i| i.next()) {
+    let addr = pick_addr(addr);
+    let listener = new_gotham_listener(addr);
+
+    let protocol = Arc::new(Http::new());
+    let new_handler = Arc::new(new_handler);
+
+    for _ in 0..threads - 1 {
+        let listener = listener.clone();
+        let protocol = protocol.clone();
+        let new_handler = new_handler.clone();
+        thread::spawn(move || run_and_serve(listener, protocol, new_handler));
+    }
+
+    run_and_serve(listener, protocol, new_handler);
+}
+
+fn run_and_serve<'a, G, NH>(listener: G, protocol: Arc<Http>, new_handler: Arc<NH>)
+where
+    G: GothamListener,
+    NH: NewHandler + 'static,
+{
+    let mut core = Core::new().expect("unable to spawn tokio reactor");
+    let handle = core.handle();
+
+    serve(listener, protocol, new_handler, handle);
+
+    core.run::<future::FutureResult<(), ()>>(future::ok(()))
+        .expect("unable to run reactor over listener");
+}
```

Comment:
This is returning immediately for me, so an application exits immediately without serving any requests. I suspect what's going on here is that the event loop only runs until the `future::ok(())` completes (immediately) and then terminates.

I think `pub fn serve` will need to return a `Future<Item = ()>` to be passed to `core.run()` so that it continues to process connections.

e.g. The way it works in the current `master` branch:
https://github.com/gotham-rs/gotham/blob/f8d03d6e879a071b7e7d341de41af0b4fe0a3710/gotham/src/os/unix.rs#L48-49